### PR TITLE
Draft: Version 2 refactor, breaking changes but much better plugin and API

### LINKED ISSUE #60: Create a parser to convert Vimscript keymaps into legendary tables
Theoretically we should be able to parse a Vimscript keymap in the form:

```VimL
vnoremap <leader>t :DoSomethingWithLeaderT<CR>
```

and convert that into the equivalent legendary keymap table:

```lua
{
  '<leader>t',
  ':DoSomethingWithLeaderT<CR>',
  mode = 'v',
  opts = {
    noremap = true,
  },
}
```

We would need to take into consideration all the different options and types of mapping commands in order to set all the right options on the table.

### PR DESCRIPTION
**Pull request by mrjones2014**
## How to Test

1. Ensure all your keymaps/commands/autocmds/functions still work (after updating your config with any API changes that might be needed)
2. Ensure documentation is accurate for new APIs

## Testing for Regressions

I have tested the following:

- [x] Triggering keymaps from `legendary.nvim` in all modes (normal, insert, visual)
- [x] Creating keymaps via `legendary.nvim`, then triggering via the keymap in all modes (normal, insert, visual)
- [x] Triggering commands from `legendary.nvim` in all modes (normal, insert, visual)
- [x] Creating commands via `legendary.nvim`, then running the command manually from the command line
- [x] `augroup`/`autocmd`s created through `legendary.nvim` work correctly

Also fixes #60 

# Release Notes

## New Features

- Option to put scratchpad in a floating window
- Command to toggle the scratchpad open and closed (mostly useful if you're using a floating window)
- Ability to define command which accept arguments
- Sorting most recently used items to the top (previously just the single most recent would be at the top)
- Parser to convert Vimscript mapping commands (e.g. `vnoremap <silent> <leader>f :SomeCommand<CR>`) to `legendary.nvim` keymaps
- Based on my testing with [`:StartupTime`](https://github.com/dstein64/vim-startuptime) and [`:PackerProfile`](https://github.com/wbthomason/packer.nvim#profiling), its about twice as fast

## Breaking Changes

- Some config options have moved, see `README.md`
- The `require('legendary.helpers')` module was moved to `require('legendary.toolbox')` and has additional utilities
- The which-key integration functions are no longer on the top-level `require('legendary')`, those are at `require('legendary.integrations.which-key')`, however if `config.which_key.auto_register == true`, they will still be auto-registered
- `kind` field was removed, instead you can use the following functions: 
  - `require('legendary.toolbox').is_keymap()`
  - `require('legendary.toolbox').is_command()`
  - `require('legendary.toolbox').is_func()`
  - `require('legendary.toolbox').is_autocmd()`
  - `require('legendary.toolbox').is_augroup()`
  - `require('legendary.toolbox').is_augroup_or_autocmd()`
- The following functions were renamed:
  - `require('legendary').bind_keymap()` => `require('legendary').keymap()`
  - `require('legendary').bind_keymaps()` => `require('legendary').keymaps()`
  - `require('legendary').bind_command()` => `require('legendary').command()`
  - `require('legendary').bind_commands()` => `require('legendary').commands()`
  - `require('legendary').bind_function()` => `require('legendary').func()`
  - `require('legendary').bind_functions()` => `require('legendary').funcs()`
  - `require('legendary').bind_autocmds()`
    - Split into two functions, `require('legendary').autocmd()` and `require('legendary').autocmds()`
 - Callbacks functions are no longer passed marks as arguments -- this behavior prevented the ability to create commands which accepted arguments themselves, as they would not have been passed to the callback function. Instead, you may use the following utilities:
   - `require('legendary.toolbox').is_visual_mode()`
   - `require('legendary.toolbox').get_marks()`
- Commands are now executed from the picker via `vim.cmd` instead of `vim.api.nvim_feedkeys`; for visual mode mappings with command string implementations, you may need to add marks to your command, for example `: '<,'>SomeCommand<CR>`

### CODE REVIEW

**Comment by mrjones2014** [comment]
@olimorris would you mind using this branch for a while and helping me file issues for anything that doesn't work and any documentation that is not accurate on this branch?

I think I've covered everything documentation-wise but it's hard to be sure.

Note that there are some breaking changes on this branch, I'll do my best to summarize them:

- Some config options have moved, see new `config.lua` or the README.md changes regarding config in this PR
- The top-level API functions have changes like `bind_keymaps()` changed to just `keymaps()`, others following the same patterns
- The helpers have moved to `legendary.toolbox`
- `kind` was removed, instead you can use `require('legendary.toolbox').is_keymap()` (there's a function for all the types)
- the `which-key` integration functions are no longer on the top-level `require('legendary')`, those are at `require('legendary.integrations.which-key')`, however if `config.which_key.auto_register == true`, they will still be auto-registered

**Comment by mrjones2014** [comment]
These changes cut my `legendary.nvim` startup time basically in half too ❤️ 

I hope it does the same for you, I'm curious if you'll see the same results.

File: lua/legendary/init.lua [hunk 1]
```diff
@@ -1,141 +1,164 @@
...
+        .. 'https://github.com/mrjones2014/legendary.nvim for more details.'
+    )
+  end, 500)
+end
 
+if vim.v.vim_did_enter then
+  show_notice()
+else
+  vim.api.nvim_create_autocmd('VimEnter', { callback = show_notice, once = true })
+end
```
Comment (mrjones2014) [inline lua/legendary/init.lua:28]: Also can you think of a better way to do this? Ideally it should be something you can acknowledge and dismiss permanently.

**Comment by olimorris** [comment]
Alright. A pretty smooth move as you can see by 935e2b3e62f141a186915bc22ce121012cb056f4. Only issue was I had actually messed up an autocmd. The old version of Legendary seemed to be forgiving of that. The new version cried and I was able to figure it out.

One thing I've noticed, I get a lot of LSP duplicates now:

Screen Shot 2022-11-04 at 09 34 43@2x

I'd even borrowed your helper function:

```lua
  if
    #vim.tbl_filter(
      function(keymap) return (keymap.desc or ""):lower() == "lsp: find references" end,
      vim.api.nvim_buf_get_keymap(bufnr, "n")
    ) > 0
  then
    return {}
  end
```

**Comment by mrjones2014** [comment]
LSP duplicates should be resolved now, the issue was I forgot to filter by buffer by default. It's now only showing items that either don't have any `buffer` option set, or have the `buffer` option set to current buffer.

**Comment by mrjones2014** [comment]
Also curious if your startup time is better 👀 

If you've got packer.nvim profiling enabled you can run `:PackerProfile` to get data like this:

image

**Comment by mrjones2014** [comment]
Ah, I took a look at your commit in your dotfiles repo -- I'm not sure how that autocmd layout worked before either actually 😂

The augroups are supposed to be a _list_ of autocmds. Looks like you've got it worked out now though. The current behavior is the intended behavior.

**Comment by olimorris** [comment]
> Ah, I took a look at your commit in your dotfiles repo -- I'm not sure how that autocmd layout worked before either actually 😂
> 
> 
> 
> The augroups are supposed to be a _list_ of autocmds. Looks like you've got it worked out now though. The current behavior is the intended behavior.

Hahaha. How that ended up in that state, I don't know. #LateNightDotfileEditing

Shall test the startup time later too.

**Comment by olimorris** [comment]
What on earth have you done to it?! Speed increase is incredible:

Version 2:
```
Oli.plugins.legendar   3.26    1.97 ▊
legendary              0.69    0.42 ▏
legendary.api.cmds     0.25    0.15
legendary.data.state   0.21    0.12
legendary.data.iteml   0.17    0.10
legendary.ui.scratch   0.15    0.09
legendary.ui           0.12    0.07
legendary.lua          0.10    0.06
legendary.data.augro   0.09    0.06
legendary.api.cache    0.06    0.03
```

Version 1:
```
Oli.plugins.legendar   7.98    4.72 █▊
legendary              4.19    2.48 ▉
legendary.bindings     2.39    1.42 ▌
legendary.cmds         1.14    0.68 ▎
legendary.config       0.61    0.36 ▏
legendary.compat.whi   0.60    0.36 ▏
legendary.utils        0.59    0.35 ▏
legendary.formatter    0.57    0.34 ▏
legendary.types        0.57    0.33 ▏
legendary.scratchpad   0.56    0.33 ▏
legendary.helpers      0.56    0.33 ▏
```

👏🏼 👏🏼 👏🏼

**Comment by olimorris** [comment]
> LSP duplicates should be resolved now, the issue was I forgot to filter by buffer by default. It's now only showing items that either don't have any `buffer` option set, or have the `buffer` option set to current buffer.

I'm still getting duplicates btw

**Comment by mrjones2014** [comment]
> What on earth have you done to it?! Speed increase is incredible

Different architecture. Keymaps are now parsed based on a schema upfront and then stored internally in a very simple format, basically tables like, for keymaps for example:

```lua
{
  -- your actual keys
  keys = string,
  description = string,
  -- implementation is converted to mode_mappings
  -- even if there's only 1, makes processing a lot
  -- simpler and faster
  mode_mappings = {
    n = { implementation = string|function, opts = table|nil }
    v = { implementation = string|function, opts = table|nil }
    x = { implementation = string|function, opts = table|nil }
    c = { implementation = string|function, opts = table|nil }
    s = { implementation = string|function, opts = table|nil }
    t = { implementation = string|function, opts = table|nil }
    i = { implementation = string|function, opts = table|nil }
  }
}
```

and for commands:

```lua
{
  cmd = string,
  implementation = function|string|nil,
  description = string,
  opts = table,
  unfinished = boolean,
}
```

This makes the initial processing a lot less complex, and a lot faster, and specifically makes _validation_ a lot simpler, enough so that I'm just using `vim.validate` now.

> I'm still getting duplicates btw

I'll check out your config and work on this now

**Comment by mrjones2014** [comment]
Oh interesting, I'm getting duplicates too. I didn't realize.

**Comment by mrjones2014** [comment]
lol. I was checking for `buf` but the actual option is `buffer` 🙈 

are you still getting duplicates now after latest commit?

File: lua/legendary/init.lua (hunk 1 above)
Comment (mrjones2014) [reply lua/legendary/init.lua:27]: Aight this shit is annoying me already. I gotta find a better solution for this. Just going to remove it for now.

**Comment by mrjones2014** [comment]
>Any plans to change the visual selection part anytime soon?

Assuming you mean what we talked about in the GitHub discussion. Does this solve your problem, or no?

```lua
  if require('legendary.toolbox').is_visual_mode() then
    -- do visual mode stuff
  else
    -- do normal mode stuff
  end
```

or alternatively, using per-mode mappings?

**Comment by mrjones2014** [comment]
>So that's complete parity of features compared to V1

A couple new features too 😉 

Namely:

- Option to put scratchpad in a floating window
- Command to toggle the scratchpad open and closed (mostly useful if you're using a floating window)
- Ability to define command which accept arguments
- Sorting most recently used items to the top (previously just the single most recent would be at the top)

**Comment by olimorris** [comment]
> > Any plans to change the visual selection part anytime soon?
> 
> Assuming you mean what we talked about in the GitHub discussion. Does this solve your problem, or no?
> 
> ```lua
>   if require('legendary.toolbox').is_visual_mode() then
>     -- do visual mode stuff
>     -- there is also this if you need it
>     local marks = require('legendary.toolbox').get_marks()
>   else
>     -- do normal mode stuff
>   end
> ```
> 
> or alternatively, using per-mode mappings?

No this is great. I've now updated for this.

**Comment by mrjones2014** [comment]
@olimorris I’d like to get some input on the best way to communicate these changes to users, here are some ideas I came up with:

- Just merge the breaking changes, users will check GitHub where I’ll have the pinned issue for breaking changes (definitely not ideal, I’d prefer to be more gently and considerate to users)
- Keep the `bind_*` function variants, and delegating internally to the new versions, and display a message via `vim.deprecate`, tag v2.0.0, then remove them and tag v2.0.1. The message will instruct users to either pin to v1.0.0 tag (which I’ll create before merging this PR), or update to v2-compatible configs
- Something else?

I don’t have a Reddit account anymore so I can’t post on the subreddit, unfortunately I think that’s where the most people would see it.

**Comment by olimorris** [comment]
Speaking from experience, I can definitely recommend adding deprecation notices. I think people are fine with being pressured (by way of messages when they start Neovim) into updating their config to a newer api, so long as their config isn't broken. That feels like the plugin author has actively made a decision to burn their config (🔥) without their permission. I've been hugely guilty of that in the past (😿).

I defo think Reddit is the place to show case the plugin again. The crowd can be very hit and miss but it pulls in the numbers! I estimate I received about 30+ stars from my last Reddit post and I lost tracked of total clones.

**Comment by mrjones2014** [comment]
Agreed 100%.

Went ahead and added deprecation notices in [5d46c0c](https://github.com/mrjones2014/legendary.nvim/pull/191/commits/5d46c0cc084dd3c571d9b5afefd7e837b28b250c) and did my best to remap things to their new equivalents.

File: lua/legendary/data/keymap.lua [hunk 2]
```diff
@@ -80,6 +80,23 @@ function Keymap:parse(tbl) -- luacheck: no unused
...
+  -- TODO remove deprecation check
+  -- check if any mapping functions are expecting arguments,
+  -- since we don't pass them anymore
+  for mode, mapping in pairs(instance.mode_mappings) do
+    if Toolbox.is_visual_mode(mode) and type(mapping.implementation) == 'function' then
+      local dbg = debug.getinfo(mapping.implementation)
+      if dbg and dbg.nparams > 0 then
+        vim.deprecate(
+          'visual_selection marks passed to keymap callbacks',
+          "require('legendary.toolbos').is_visual_mode() and require('legendary.toolbox').get_marks()",
```
Comment (olimorris) [inline lua/legendary/data/keymap.lua:92]: Slight typo on line 92 btw

**Comment by olimorris** [comment]
Also, just received this error after updating:

Screen Shot 2022-11-08 at 08 35 57@2x

Could this be handled gracefully?

**Comment by mrjones2014** [comment]
That message is using `vim.deprecate` and it is copying the value to the correct config value in the new setup.

How would you suggest I handle it more gracefully?

**Comment by mrjones2014** [comment]
Nice, I like that a lot. Updated.

**Comment by mrjones2014** [comment]
I think I wanna re-record the demo gif in the README.md and then this PR is lookin pretty good 😎

**Comment by mrjones2014** [comment]
Planning to release this to master tomorrow some time, probably in the morning. Trying to decide if I should make a Reddit account just to post an update there lol

**Comment by olimorris** [comment]
Just read the new README and had some thoughts (feel free to discard every single one of them):

In Safari, I find that `mp4` videos do not autoplay in GitHub, whilst they do in Chrome. Perhaps that is an issue on my end but worth testing to make sure. Because if I come to `Legendary` for the first time, I'm not sure I would totally understand what we mean by a Legend (especially if English is not a primary language). In VS Code I see `Cmd+Shift+P` being called the Command Palette. Is there a way to rephrase the introductory sentence to be more explicit about what `Legendary` does? Something crude like:

"Define your keymaps, commands, and autocommands as simple Lua tables, building a legend (or command palette) at the same time."

I have a bit of a dislike for videos as the main media in a README as well because it requires me to watch 10+ seconds of someones plugin to try and workout what it does (try watching the `leap.nvim` one). Videos are absolutely useful for demonstrating features once the reader has grasped the elements of the plugin but upfront, I'd love a nice logo incorporating a few snapshots of `Legendary`.

**Comment by mrjones2014** [comment]
>In Safari, I find that mp4 videos do not autoplay in GitHub, whilst they do in Chrome

Yeah, I was having a hard time keeping the `gif` file size small enough. I can give it another shot as a gif.

As far as the tagline, I can do something like: "Define your keymaps, commands, and autocommands as simple Lua tables, building a legend (like VS Code's Command Palette) at the same time."

Any specific thoughts on what a logo should look like? I had played with the idea but I'm bad at designing stuff like that 😂

**Comment by mrjones2014** [comment]
Okay, got a gif working with appropriate file size.

**Comment by olimorris** [comment]
> As far as the tagline, I can do something like: "Define your keymaps, commands, and autocommands as simple Lua tables, building a legend (like VS Code's Command Palette) at the same time."

Personally, I think anyone new to the plugin will grasp what it does rig away. Also, worth updating `awesome-neovim` with the new tagline?

> Any specific thoughts on what a logo should look like? I had played with the idea but I'm bad at designing stuff like that 😂 

Haha I should not be giving any advice out on this 🤣. For my `tmux-pomodoro-plus` plugin I found a cool stock image and stuck it over the tmux logo and made it a banner. Could always be an option in place of just a cool font and some screenshots placed together in a banner.

**Comment by mrjones2014** [comment]
>Also, worth updating awesome-neovim with the new tagline?

Good call, PR: https://github.com/rockerBOO/awesome-neovim/pull/599

**Comment by mrjones2014** [comment]
reddit post: https://www.reddit.com/r/neovim/comments/yqiefg/legendarynvim_now_much_faster_with_new_features/

**Comment by olimorris** [comment]
Congratulations @mrjones2014! A huge milestone.

**Comment by mrjones2014** [comment]
Thanks! I'm really happy with the way it turned out, and tbh I'm happy to be free of Teal. Rewriting to Teal about 6 months ago _really_ did not have the payoff I'd hoped. Sure, static types are nice, but using EmmyLua annotation comments with the Lua language server is close enough and the tooling surrounding Teal (LSP, editor support, compiler, etc.) are pretty sub-par developer experience IMHO.

**Comment by mrjones2014** [comment]
It will be so so so much easier to reason about when fixing bugs and adding new features now as well. The architecture before was getting to be a bit of a mess.

**Pull request merged**